### PR TITLE
[FEATURE] Borg commands accept absolute pathes, #4029

### PR DESCRIPTION
**Pull request by fantasya-pbem**
Remove all leading slashes in patterns of all sorts but re: style.

This change will break some tests, but as far as I could determine it has no other side effects. I left out changing the re: patterns because removing the leading slash can invalidate all non-trivial patterns (like /?some/dir) or is complicated (like ^/some/dir).

I checked ``extract`` and ``mount`` to verify that all --exclude pattern styles work and the pathes args are recognized.

What's your opinion? Acceptable, or too esoteric?

### CODE REVIEW

**Comment by fantasya-pbem** [comment]
In a second step I fixed most tests.

**Comment by ThomasWaldmann** [comment]
there are some formatting issues which are hopefully also found by the flake8 checker:

https://travis-ci.org/github/borgbackup/borg/jobs/673750516

**Comment by ThomasWaldmann** [comment]
Did you think about how this relates to native windows support?

**Comment by fantasya-pbem** [comment]
> Did you think about how this relates to native windows support?

No, and I cannot run tests or try the changes on Windows. I guess Borg also saves only relative paths, just that they have a backslash in them? ``\Programs\MyTool`` becomes ``Programs\MyTool``, and ``C:\some\file`` becomes ``some\file``? So this fix would work when not using drive letters I guess, isn't it?

**Comment by ThomasWaldmann** [comment]
I didn't review your changes yet. If you only deal with pathes as seen in archives, there is never a leading /. But on pathes that somehow come from the user (cli args or other) might have leading / (or \).
See also the other PRs.

**Comment by ThomasWaldmann** [comment]
```
flake8 run-test: commands[0] | flake8 src scripts conftest.py
src/borg/patterns.py:218:69: E703 statement ends with a semicolon
src/borg/patterns.py:218:70: E261 at least two spaces before inline comment
src/borg/patterns.py:258:51: E261 at least two spaces before inline comment
src/borg/patterns.py:282:43: E261 at least two spaces before inline comment
src/borg/patterns.py:295:31: E261 at least two spaces before inline comment
```

File: src/borg/patterns.py [hunk 1]
```diff
@@ -235,7 +235,9 @@ class PathPrefixPattern(PatternBase):
     PREFIX = "pp"
 
     def _prepare(self, pattern):
-        self.pattern = os.path.normpath(pattern).rstrip(os.path.sep) + os.path.sep
+        sep = os.path.sep
+
+        self.pattern = (os.path.normpath(pattern).rstrip(sep) + sep).lstrip(sep)  # sep at beginning is removed
```
Comment (ThomasWaldmann) [inline src/borg/patterns.py:240]: how about:
- normpath
- lstrip
- rstrip
- add sep

would not require parens like seen above.

**Comment by fantasya-pbem** [comment]
> did you check whether we test all the use cases where the matcher is used, like create, extract, mount, ...?

Yes, I did a search for the PatternMatcher class and found that these changes here should cover all occurrences of paths and patterns. As the matcher classes are not used for other stuff, I think these few changed lines are the minimum solution.

File: src/borg/patterns.py (hunk 1 above)
Comment (ThomasWaldmann) [reply src/borg/patterns.py:240]: ah, ok, I overlooked that case.

**Comment by ThomasWaldmann** [comment]
Thanks!

**Pull request merged**